Proposed changes for #6

sorear opened the pull request:
Not yet tested.

ddevault commented:
Note, I updated staging to the latest musl release and did some rebasing. I rebased your PR against those changes as well.

sorear committed:
riscv64: jmp_buf reorganization

1. fs3 is no longer saved and restored twice. This saves 8 bytes in all
modes.

2. The size of __jmp_buf (not jmp_buf) has been corrected to match the
size of the non-signal registers saved and restored.

The total size of jmp_buf now matches glibc; the fields are still in a
different order.

3. Additional information for sigsetjmp is now stored in words 0 and 2
after __jmp_buf, matching the address usage in aarch64.

sorear committed:
riscv32: jmp_buf reorganization

1. fs3 is no longer saved and restored twice. This saves 8 bytes in all
modes.

2. The size of __jmp_buf (not jmp_buf) has been corrected to match the
size of the non-signal registers saved and restored.

The total size of jmp_buf now matches glibc; the fields are still in a
different order.

3. Additional information for sigsetjmp is now stored in words 0 and 2
after __jmp_buf, matching the address usage in aarch64.

sorear committed:
riscv64: Scrub usage of j/jal

j and jal with GNU binutils generate instructions and relocations only
usable for a range of +- 1 MiB.  RISC-V musl ld.so is currently half of
that, but this could be a problem for static linking.

Replace with tail and call, pseudoinstructions that have +- 2 GiB range;
with relaxation enabled the same opcodes will be generated for small
binaries.

sorear committed:
riscv32: Scrub use of j/jal

j and jal with GNU binutils generate instructions and relocations only
usable for a range of +- 1 MiB.  RISC-V musl ld.so is currently half of
that, but this could be a problem for static linking.

Replace with tail and call, pseudoinstructions that have +- 2 GiB range;
with relaxation enabled the same opcodes will be generated for small
binaries.

sorear committed:
riscv64: simplify longjmp return value assignment

sorear committed:
riscv32: simplify longjmp return value assignment

sorear committed:
riscv64: s/scall/ecall/g

The instruction was renamed in RISC-V 2.0 (2014) and binutils'
acceptance of the old name is not a documented or guaranteed behavior.

sorear committed:
s/scall/ecall/g

The instruction was renamed in RISC-V 2.0 (2014) and binutils'
acceptance of the old name is not a documented or guaranteed behavior.

ddevault commented:
Tested this and it looks good. Thanks!
